### PR TITLE
Add JSON Sink for Logs, Add output dir command line flag

### LINKED ISSUE #353: Add JSON log sink for Windows client
Should format in GELF (graylog extended log format probably). Also consider making a syslog log sink

### PR DESCRIPTION
**Pull request by ION28**
Closes #353 . Server/Network sink will be addressed in a separate issue.

Output directory option was requested by jrstore in Discord

### CODE REVIEW

File: BLUESPAWN-win-client/src/util/processes/ParseCobalt.cpp [hunk 1]
```diff
@@ -2,243 +2,237 @@
...
+                if(!wrapper) {}
+                for(size_t idx1 = 0; idx1 < wrapper.GetSize(); idx1++) {
+                    if(dwRawDword == 0x69696969) {
+                        wrapper.SetByte(idx1, wrapper[idx1] ^ 0x69);
+                    } else {
+                        wrapper.SetByte(idx1, wrapper[idx1] ^ 0x2E);
+                    }
+                }
+                return wrapper;
+            } else if(dwRawDword == 0x2E2E2E2E || dwRawDword == 0x69696969) {
```
Comment (ION28) [inline BLUESPAWN-win-client/src/util/processes/ParseCobalt.cpp:38]: Codacy found an issue: [Condition 'dwRawDword==774778414' is always false](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=6145581)

File: BLUESPAWN-win-client/src/util/processes/ParseCobalt.cpp [hunk 2]
```diff
@@ -2,243 +2,237 @@
...
+                auto wrapper{
+                    memory.GetOffset(section->PointerToRawData + 0x30).Convert<DWORD>().ToAllocationWrapper(4096)
+                };
+                for(size_t idx1 = 0; idx1 < wrapper.GetSize(); idx1++) {
+                    if(dwRawDword == 0x69696969) {
```
Comment (ION28) [inline BLUESPAWN-win-client/src/util/processes/ParseCobalt.cpp:43]: Codacy found an issue: [Condition 'dwRawDword==1768515945' is always false](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=6145581)

File: BLUESPAWN-win-client/src/util/processes/ParseCobalt.cpp (hunk 1 above)
Comment (ION28) [inline BLUESPAWN-win-client/src/util/processes/ParseCobalt.cpp:38]: Codacy found an issue: [Condition 'dwRawDword==774778414' is always false](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=6145581)

File: BLUESPAWN-win-client/src/util/processes/ParseCobalt.cpp (hunk 2 above)
Comment (ION28) [inline BLUESPAWN-win-client/src/util/processes/ParseCobalt.cpp:43]: Codacy found an issue: [Condition 'dwRawDword==1768515945' is always false](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=6145581)

File: BLUESPAWN-win-client/src/util/log/JSONSink.cpp [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        ResumeThread(thread);
+    }
+
+    JSONSink::JSONSink(const std::wstring& wOutputDir) :
+        thread{ CreateThread(nullptr, 0, PTHREAD_START_ROUTINE(UpdateLog), this, 0, nullptr) } {
+        SYSTEMTIME time{};
+        GetLocalTime(&time);
+        wFileName = wOutputDir + L"\\bluespawn-" + ToWstringPad(time.wMonth) + L"-" + ToWstringPad(time.wDay) + L"-" +
+                    ToWstringPad(time.wYear, 4) + L"-" + ToWstringPad(time.wHour) + ToWstringPad(time.wMinute) + L"-" +
+                    ToWstringPad(time.wSecond) + L".xml";
```
Comment (Jack-McDowell) [inline BLUESPAWN-win-client/src/util/log/JSONSink.cpp:42]: nope

File: BLUESPAWN-win-client/src/util/log/JSONSink.cpp [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+    void JSONSink::AddAssociation(IN DWORD detection_id, IN DWORD associated, IN double strength) {
+        for(auto& item : JSONDoc["bluespawn"]["detections"]) {
+            if(item["id"] == detection_id) {
+                if(item.find("associated-detections") != item.end()) {
+                    /// update association strength for association already in associated-detections of detection_id
+                    for(auto& child : item["associated-detections"]) {
+                        if(child["id"] == associated) {
+                            const auto cur_strength = child.find("strength");
+                            child.erase(cur_strength);
```
Comment (Jack-McDowell) [inline BLUESPAWN-win-client/src/util/log/JSONSink.cpp:95]: Acquire the appropriate critical sections first.

File: BLUESPAWN-win-client/src/util/log/JSONSink.cpp [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+            JSONDoc["bluespawn"]["log-messages"].push_back(msg);
+        }
+    }
+
+    bool JSONSink::operator==(const LogSink& sink) const {
+        return (bool) dynamic_cast<const JSONSink*>(&sink) &&
+               dynamic_cast<const JSONSink*>(&sink)->wFileName == wFileName;
+    }
+
+    void JSONSink::Flush() {
```
Comment (Jack-McDowell) [inline BLUESPAWN-win-client/src/util/log/JSONSink.cpp:235]: Needs to be regularly called during monitor mode

**Review by Jack-McDowell — CHANGES_REQUESTED** [review]

File: BLUESPAWN-win-client/src/util/log/JSONSink.cpp (hunk 3 above)
Comment (ION28) [reply BLUESPAWN-win-client/src/util/log/JSONSink.cpp:42]: good catch

File: BLUESPAWN-win-client/src/util/log/JSONSink.cpp (hunk 5 above)
Comment (ION28) [reply BLUESPAWN-win-client/src/util/log/JSONSink.cpp:235]: my code here looks the same as what's in XMLSink?

Comment (Jack-McDowell) [reply BLUESPAWN-win-client/src/util/log/JSONSink.cpp:237]: Oh, I might have missed it last time, but you needed the thread watching for the flushlogs event and calling this. You have it now, at least.

File: BLUESPAWN-win-client/src/util/log/JSONSink.cpp (hunk 4 above)
Comment (Jack-McDowell) [reply BLUESPAWN-win-client/src/util/log/JSONSink.cpp:97]: My bad, it was good before

**Review by Jack-McDowell — APPROVED** [review]

**Pull request merged**